src/diffusion_thomas_algorithm.cc

```diff
@@ -0,0 +1,354 @@
+  // NOLINTBEGIN(cppcoreguidelines-pro-bounds-pointer-arithmetic)
+  const auto* all_concentrations = GetAllConcentrations();
+  const real_t initial_concentration = all_concentrations[ind];
+  SetConcentration(ind, initial_concentration / thomas_denom[0]);
+
+  // Forward elimination loop
+  for (unsigned int inner = 1; inner < resolution_; inner++) {
+    ind = GetLoopIndex(direction, outer, middle, inner);
+    const real_t current_concentration = all_concentrations[ind];
+    const real_t prev_concentration = all_concentrations[ind - jump];
+    SetConcentration(ind,
+                     (current_concentration + constant1_ * prev_concentration) /
+                         thomas_denom[inner]);
+  }
+  // NOLINTEND(cppcoreguidelines-pro-bounds-pointer-arithmetic)
+}
+
+void DiffusionThomasAlgorithm::BackSubstitution(
+    unsigned int direction, unsigned int outer, unsigned int middle,
+    const std::vector<real_t>& thomas_c, unsigned int jump) {
+  // NOLINTBEGIN(cppcoreguidelines-pro-bounds-pointer-arithmetic)
+  const auto* all_concentrations = GetAllConcentrations();
+
+  // Back substitution loop
+  for (int inner = static_cast<int>(resolution_) - 2; inner >= 0; inner--) {
+    const size_t ind = GetLoopIndex(direction, outer, middle,
+                                    static_cast<unsigned int>(inner));
+    const real_t current_concentration = all_concentrations[ind];
+    const real_t next_concentration = all_concentrations[ind + jump];
+    SetConcentration(
+        ind, current_concentration -
+                 thomas_c[static_cast<size_t>(inner)] * next_concentration);
+  }
+  // NOLINTEND(cppcoreguidelines-pro-bounds-pointer-arithmetic)
+}
+
+size_t DiffusionThomasAlgorithm::GetLoopIndex(unsigned int direction,
+                                              unsigned int outer,
+                                              unsigned int middle,
+                                              unsigned int inner) const {
+  switch (direction) {
+    case 0:  // X-direction: outer=k, middle=j, inner=i
+      return GetBoxIndex(inner, middle, outer);
+    case 1:  // Y-direction: outer=k, middle=i, inner=j
+      return GetBoxIndex(middle, inner, outer);
+    case 2:  // Z-direction: outer=j, middle=i, inner=k
+      return GetBoxIndex(middle, outer, inner);
+    default:
+      return 0;
+  }
+}
+
+void DiffusionThomasAlgorithm::ComputeConsumptionsSecretions() {
+  // This method is called to compute the consumptions and secretions of
+  // substances by the tumor cells. It iterates over all agents and applies the
+  // consumption and secretion behaviors defined in the TumorCell class.
+  auto* rm = bdm::Simulation::GetActive()->GetResourceManager();
+  // in a future version of BioDynaMo this should be parallelized getting the
+  // agents inside each chemical voxel and treating each voxel independently.
+  rm->ForEachAgent([this](bdm::Agent* agent) {
+    if (auto* cell = dynamic_cast<TumorCell*>(agent)) {
```

Comment:
Can we add a fixme saying this dynamic_cast here is slow. Maybe we can ping Lukas and tell him that we could implement an LLVM-style rtti that's fast for next iteration...